src/libdrakvuf/win-processes.c

```diff
@@ -164,3 +170,107 @@ char *drakvuf_get_process_name(drakvuf_t drakvuf, addr_t eprocess_base) {
+                                    uint32_t *thread_id )
+{
+    addr_t p_tid ;
+    addr_t ethread = drakvuf_get_current_thread( drakvuf, vcpu_id, regs );
+
+    if ( ethread )
+    {
+        if ( vmi_read_addr_va( drakvuf->vmi, ethread + offsets[ ETHREAD_CID ] + offsets[ CLIENT_ID_UNIQUETHREAD ],
+                               0,
+                               &p_tid ) == VMI_SUCCESS )
+        {
+            *thread_id = p_tid;
+
+            return true;
+        }
+    }
+
+    return false ;
+}
+
+
+/////////////////////////////////////////////////////////////////////////////////////////////
+
+// Microsoft PreviousMode KTHREAD explanation:
+// https://msdn.microsoft.com/en-us/library/windows/hardware/ff559860(v=vs.85).aspx
+
+bool drakvuf_get_thread_previous_mode( drakvuf_t drakvuf, addr_t kthread, privilege_mode_t *previous_mode )
+{
+    if ( kthread )
+    {
+        if ( vmi_read_8_va( drakvuf->vmi, kthread + offsets[ KTHREAD_PREVIOUSMODE ], 0,
+                            (uint8_t *)previous_mode ) == VMI_SUCCESS )
+        {
+            if ( ( *previous_mode == KERNEL_MODE ) || ( *previous_mode == USER_MODE ) )
+
+                return true ;
+        }
+    }
+
+    return false ;
+}
+
+bool drakvuf_get_current_thread_previous_mode( drakvuf_t drakvuf, drakvuf_trap_info_t *info, 
+                                               privilege_mode_t *previous_mode )
+{
+    addr_t kthread = drakvuf_get_current_thread( drakvuf, info->vcpu, info->regs );
+
+    return drakvuf_get_thread_previous_mode( drakvuf, kthread, previous_mode );
+}
+
+
+/////////////////////////////////////////////////////////////////////////////////////////////
+
+
+bool drakvuf_is_ethread( drakvuf_t drakvuf, drakvuf_trap_info_t *info, addr_t ethread_addr )
+{
+    dispatcher_object_t dispatcher_type ;
+    access_context_t ctx = {
+            .translate_mechanism = VMI_TM_PROCESS_DTB,
+            .dtb = info->regs->cr3,
+    };
+
+    ctx.addr = ethread_addr + offsets[ ETHREAD_TCB ] + offsets[ KTHREAD_HEADER ]
+                            + offsets[ DISPATCHER_TYPE ] ;
+
+    if ( vmi_read_8( drakvuf->vmi, &ctx, (uint8_t *)&dispatcher_type ) == VMI_SUCCESS )
+    {
+        if ( dispatcher_type == DISPATCHER_THREAD_OBJECT )
+
+            return true ;
+    }
+
+    return false ;
+}
+
+
+/////////////////////////////////////////////////////////////////////////////////////////////
+
+
+bool drakvuf_is_eprocess( drakvuf_t drakvuf, drakvuf_trap_info_t *info, addr_t eprocess_addr )
+{
+    dispatcher_object_t dispatcher_type ;
+    access_context_t ctx = {
+            .translate_mechanism = VMI_TM_PROCESS_DTB,
+            .dtb = info->regs->cr3,
+    };
+
+    ctx.addr = eprocess_addr + offsets[ EPROCESS_PCB ] + offsets[ KPROCESS_HEADER ]
+                             + offsets[ DISPATCHER_TYPE ] ;
+
+    if ( vmi_read_8( drakvuf->vmi, &ctx, (uint8_t *)&dispatcher_type ) == VMI_SUCCESS )
+    {
+        if ( dispatcher_type == DISPATCHER_PROCESS_OBJECT )
+
```

Comment:
I agree, it's very strange taking into account that these routines are just helper routines and executed only on demand by the plugins and, as far as I know, there are no plugins using them.